README.md
Weblogic Cookbook
=======================

This is a placeholder cookbook for the Weblogic cookbook

Scope
-----
This cookbook is for installing a basic Weblogic 11g or 12c standalone installation.

Requirements
------------
* Chef 11 or higher
* Ruby 1.9 (preferably from the Chef full-stack installer)
* Weblogic 12c installation JAR file (wls_121200.jar)

Recipes
-------
### weblogic::default

This recipe installs Weblogic 11g or 12c using the generic silent installer.

Usage
-----

### TODO

Attributes
----------

### TODO

License & Authors
-----------------
- Author:: Michael Goetz (<mpgoetz@getchef.com>)

```text
Copyright:: 2009-2014 Chef Software, Inc

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```

CHANGELOG.md
Weblogic
========

v0.2.0
-------------------
- Extend to handle 11g or 12c installations

v0.1.3 
-------------------
- Initial working recipe for installing Weblogic 12c


v0.1.1 
-------------------


v1.0.0 
-------------------
- Initial release


spec/unit/default_spec.rb
require 'spec_helper'

describe 'weblogic::default' do
  let(:chef_run) { ChefSpec::Runner.new.converge(described_recipe) }

  it 'cteates the weblogic_install response file' do
    expect(chef_run).to create_template("#{Chef::Config[:file_cache_path]}/weblogic_install.rsp")
  end

  it 'creates the silent.xml template' do
    expect(chef_run).to create_template("#{Chef::Config[:file_cache_path]}/silent.xml")
  end
 
  it 'creates the weblogic application user' do
    expect(chef_run).to create_user('oracle')
  end

  it 'creates the weblogic application group' do
    expect(chef_run).to create_group('oracle')
  end

  it 'creates the oraInst.loc file' do
    expect(chef_run).to create_template('/etc/oraInst.loc')
  end

  it 'Downloads the weblogic jar' do
    expect(chef_run).to create_remote_file("#{Chef::Config[:file_cache_path]}/wls_121200.jar")
  end

  it 'executes the weblogic jar to do the install' do
    expect(chef_run).to run_execute("java -jar #{Chef::Config[:file_cache_path]}/wls_121200.jar -silent -silent_xml=#{Chef::Config[:file_cache_path]}/silent.xml -responseFile #{Chef::Config[:file_cache_path]}/weblogic_install.rsp -Djava.security.egd=file:/dev/./urandom")
  end

end

describe "weblogic::create_domain" do
  let(:chef_run) { ChefSpec::Runner.new.converge(described_recipe) }
  domain_home= '/home/oracle/oracle_mw/domain_home'
  it 'creates domain_home dir' do
    expect(chef_run).to create_directory("#{domain_home}")
  end

  it 'creates domain_home/example_name dir' do
    expect(chef_run).to create_directory("#{domain_home}/example_name")
  end

  it 'creates the create_example_name_domain.py template' do
      expect(chef_run).to create_template("#{domain_home}/create_example_name_domain.py")
  end

  it 'creates domain dir' do
    expect(chef_run).to create_directory("#{domain_home}/example_name")
  end

  it 'executes the domain creation through wlst.sh' do
    expect(chef_run).to run_execute("/home/oracle/oracle_mw/wlserver/common/bin/wlst.sh #{domain_home}/create_example_name_domain.py")
  end

  it 'templates weblogic init script' do
      expect(chef_run).to create_template("/etc/init.d/weblogic")
  end

  it 'enables weblogic service' do
      expect(chef_run).to enable_service("weblogic")
  end 
end


recipes/create_domain.rb

domain_home_dirs = [
  node['weblogic']['domain_home'],
  "#{node['weblogic']['domain_home']}/#{node['weblogic']['domain_name']}"
]

domain_home_dirs.each do |dirname|
  directory dirname do
    owner node['weblogic']['user']
    group node['weblogic']['group']
    mode 00755
    recursive true
    action :create
  end
end

template "#{node['weblogic']['domain_home']}/create_#{node['weblogic']['domain_name']}_domain.py" do
  source 'create_domain.py.erb'
  owner node['weblogic']['user']
  group node['weblogic']['group']
  mode '0755'
end

execute "wlst.sh create_#{node['weblogic']['domain_name']}_domain.py" do
  command "#{node['weblogic']['wls_install_dir']}/common/bin/wlst.sh #{node['weblogic']['domain_home']}/create_#{node['weblogic']['domain_name']}_domain.py"
  creates "#{node['weblogic']['domain_home']}/#{node['weblogic']['domain_name']}/bin"
  user node['weblogic']['user']
  group node['weblogic']['group']
end

if node['weblogic']['install_init_d']
  template '/etc/init.d/weblogic' do
    source 'weblogic_init.d.sh.erb'
    owner 'root'
    group 'root'
    mode '0744'
  end

  service 'weblogic' do
    supports :status => true, :restart => false, :reload => false
    action [:enable, :start]
  end
end


recipes/default.rb
#
# Cookbook Name:: weblogic
# Recipe:: default
#
# Copyright 2009-2014 Chef Software, Inc
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

if node['weblogic']['install_java']
  include_recipe 'java'
end

# Create the weblogic application user
user node['weblogic']['user'] do
  action :create
end

# Create the weblogic application group
group node['weblogic']['group'] do
  members [node['weblogic']['user']]
end

# Create the oraInst.loc file
template '/etc/oraInst.loc' do
  source 'oraInst.loc.erb'
  owner 'root'
  group 'root'
  mode '0644'
end

# Create the Weblogic install response file
template "#{Chef::Config[:file_cache_path]}/weblogic_install.rsp" do
  source 'weblogic_install.rsp.erb'
  owner 'root'
  group 'root'
  mode '0644'
end

# Create the silent.xml file
template "#{Chef::Config[:file_cache_path]}/silent.xml" do
  source 'silent.xml.erb'
  owner 'root'
  group 'root'
  mode '0644'
end

if node['weblogic']['version'] == '11g'
  remote_file "#{Chef::Config[:file_cache_path]}/wls1035_generic.jar" do
    owner 'root'
    group 'root'
    mode '0755'
    source node['weblogic']['11g']['installer_download']
    checksum node['weblogic']['11g']['installer_checksum']
  end

  # Install Weblogic
  execute "Install Weblogic" do
    command "java -jar #{Chef::Config[:file_cache_path]}/wls1035_generic.jar -silent_xml=#{Chef::Config[:file_cache_path]}/silent.xml -mode=silent"
    user node['weblogic']['user']
    group node['weblogic']['group']
    creates node['weblogic']['wls_install_dir']
    action :run
  end

  #Change the directory ownership from root to oracle
  bash "chown" do
    code <<-EOH
    chown -R #{node['weblogic']['user']}:#{node['weblogic']['group']} #{node['weblogic']['beahome']}
    EOH
  end
end
end

if node['weblogic']['version'] == '12c'
  # Download the Universal Installer JAR
  remote_file "#{Chef::Config[:file_cache_path]}/wls_121200.jar" do
    owner 'root'
    group 'root'
    mode '0744'
    source node['weblogic']['12c']['installer_download']
    checksum node['weblogic']['12c']['installer_checksum']
  end

# Install Weblogic
  execute "Install Weblogic" do
    command "java -jar #{Chef::Config[:file_cache_path]}/wls_121200.jar -silent -silent_xml=#{Chef::Config[:file_cache_path]}/silent.xml -responseFile #{Chef::Config[:file_cache_path]}/weblogic_install.rsp -Djava.security.egd=file:/dev/./urandom"
    user node['weblogic']['user']
    group node['weblogic']['group']
    creates node['weblogic']['wls_install_dir']
    action :run
  end
end